lib/evaluate/real.h

```diff
@@ -0,0 +1,307 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_EVALUATE_REAL_H_
+#define FORTRAN_EVALUATE_REAL_H_
+
+#include "common.h"
+#include "integer.h"
+#include "rounding-bits.h"
+#include <cinttypes>
+#include <limits>
+
+namespace Fortran::evaluate::value {
+
+// Models IEEE binary floating-point numbers (IEEE 754-2008,
+// ISO/IEC/IEEE 60559.2011).  The first argument to this
+// class template must be (or look like) an instance of Integer<>;
+// the second specifies the number of effective bits in the fraction;
+// the third, if true, indicates that the most significant position of the
+// fraction is an implicit bit whose value is assumed to be 1 in a finite
+// normal number.
+template<typename WORD, int PRECISION, bool IMPLICIT_MSB = true> class Real {
+public:
+  using Word = WORD;
+  static constexpr int bits{Word::bits};
+  static constexpr int precision{PRECISION};
+  static constexpr bool implicitMSB{IMPLICIT_MSB};
+  static constexpr int significandBits{precision - implicitMSB};
+  static constexpr int exponentBits{bits - significandBits - 1 /*sign*/};
+  static_assert(precision > 0);
+  static_assert(exponentBits > 1);
+  static constexpr std::uint64_t maxExponent{(1 << exponentBits) - 1};
+  static constexpr std::uint64_t exponentBias{maxExponent / 2};
+
+  constexpr Real() {}  // +0.0
+  constexpr Real(const Real &) = default;
+  constexpr Real(const Word &bits) : word_{bits} {}
+  constexpr Real &operator=(const Real &) = default;
+
+  // TODO conversion from (or to?) (other) real types
+  // TODO AINT/ANINT, CEILING, FLOOR, DIM, MAX, MIN, DPROD, FRACTION
+  // HUGE, INT/NINT, MAXEXPONENT, MINEXPONENT, NEAREST, OUT_OF_RANGE,
+  // PRECISION, HUGE, TINY, RRSPACING/SPACING, SCALE, SET_EXPONENT, SIGN
+
+  constexpr bool IsNegative() const {
+    return !IsNotANumber() && word_.BTEST(bits - 1);
+  }
+  constexpr bool IsNotANumber() const {
+    return Exponent() == maxExponent && !GetSignificand().IsZero();
+  }
+  constexpr bool IsQuietNaN() const {
+    return Exponent() == maxExponent &&
+        GetSignificand().BTEST(significandBits - 1);
+  }
+  constexpr bool IsSignalingNaN() const {
+    return IsNotANumber() && !GetSignificand().BTEST(significandBits - 1);
+  }
+  constexpr bool IsInfinite() const {
+    return Exponent() == maxExponent && GetSignificand().IsZero();
+  }
+  constexpr bool IsZero() const {
+    return Exponent() == 0 && GetSignificand().IsZero();
+  }
+  constexpr bool IsDenormal() const {
+    return Exponent() == 0 && !GetSignificand().IsZero();
+  }
+
+  constexpr Real ABS() const {  // non-arithmetic, no flags returned
+    Real result;
+    result.word_ = word_.IBCLR(bits - 1);
+    return result;
+  }
```

Comment:
Why not `return Real(word_.IBCLR(bits - 1));`?

Reply:
I didn't initially have a Real() constructor that accepted a Word argument.  Will reduce.